packages/push/src/instruction/int/negate.rs

```diff
@@ -0,0 +1,132 @@
+///
+/// The `IntInstruction::Negate` instruction takes the following inputs:
+///    - `i64` stack
+///      - One value
+///
+/// # Behavior
+///
+/// The `IntInstruction::Negate` instruction negates the top value of
+/// the `i64` stack. The one exception (as described above) is when
+/// the value is `i64::MIN`, where `Negate` removes it and pushes
+/// on `i64::MAX` in its place.
+///
+/// ## Action Table
+///
+/// The table below indicates the behavior in each of the different
+/// cases.
+///
+///    - The "`i64` stack" column indicates the value of the top of the `i64`
+///      stack, or whether it exists.
+///    - The "Success" column indicates whether the instruction succeeds, and if
+///      not what kind of error is returned:
+///       - ✅: success (so two copies of the top block on the `Exec stack`)
+///       - ❗: recoverable error, with links to the error kind
+///       - ‼️: fatal error, with links to the error kind
+///    - The "Note" column briefly summarizes the action state in that case
+///
+/// | `i64`` stack  |  Success | Note |
+/// | ------------- | ------------- | ------------- |
+/// | `i64::MIN`    | ✅ | `i64::MAX` is pushed on the `i64` stack |
+/// | exists, not `i64::MIN` | ✅ | Negates top value of `i64` stack |
+/// | missing | [❗..](crate::push_vm::stack::StackError::Underflow) | State is unchanged |
+///
+/// # Errors
+///
+/// Returns a
+/// [`StackError::Underflow`](crate::push_vm::stack::StackError::Underflow)
+/// error when the `i64` stack is empty.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+pub struct Negate;
+
+impl<S> Instruction<S> for Negate
+where
+    S: Clone + HasStack<i64>,
+{
+    type Error = PushInstructionError;
+
+    fn perform(&self, state: S) -> InstructionResult<S, Self::Error> {
+        let int_stack = state.stack::<i64>();
+        int_stack
+            .top()
+            .map(|x| x.saturating_neg())
+            .replace_on(1, state)
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+mod tests {
+
+    use proptest::prop_assert_eq;
+    use test_strategy::proptest;
+
+    use crate::{
+        instruction::{Instruction, IntInstruction},
+        push_vm::{push_state::PushState, HasStack},
+    };
+
+    // We need to make sure `Negate` properly handles the
+    // case where the value being negated is `i64::MIN`.
+    // Simply negating that value will wrap around to
+    // `i64::MAX_VALUE`, which is definitely not what we want.
+    #[test]
+    fn handle_min() {
+        let input = i64::MIN;
+        let state = PushState::builder()
+            .with_max_stack_size(1)
+            .with_int_values(std::iter::once(input))
+            .unwrap()
+            .with_no_program()
+            .build();
+        let result = IntInstruction::negate().perform(state).unwrap();
+        assert_eq!(result.stack::<i64>().size(), 1);
+        assert_eq!(*result.stack::<i64>().top().unwrap(), i64::MAX);
+    }
+
+    #[proptest]
+    fn negate(#[any] x: i64) {
```

Comment:
I need to fix this `proptest` for `Negate` so that it won't fail if the value generated by `any` is `i64::MIN`.

Reply:
This is now dealt with in the following commit.